### ISSUE TITLE
send_ack hits assert when running out of frame space

### ISSUE THREAD

**Issue by sknat** [h2o/quicly#240]
Hi Kazuo, I'm back with a new assert :)

This time I'm hitting `assert(s->dst != s->dst_payload_from);` in `commit_send_packet` when running out of space in `send_ack` :

https://github.com/h2o/quicly/blob/98c8f2d0883e2282d55d19fb5a2e7cbbf59495a7/lib/quicly.c#L2524

I think `allocate_frame` leaves the send context in a state where `s->dst == s->dst_payload_from` and as we don't put anything in the ack frame if we realize we don't have enough space for everything, we hit the assert.

I'm working on a patch : for now checking in advance that we have the space and not relying on `quicly_encode_ack_frame` to tell us. But I don't know if it's ideal. Also I seem to be hitting more asserts... But I did reach the 100k streams this time !

Thanks in advance for your help

**Comment by kazuho** [comment]
Thank you for reporting the issue.

Am I correct in assuming that we hit this issue when the endpoint is sending large amount of data (while the peer is sending nothing), then it receives something from the peer?

Assuming that is the case, I think this is part of #216. I have created a table discussing possible countermeasures that we need to implement https://github.com/h2o/quicly/issues/216#issuecomment-562552461.

In short, there should be a limit for number of gaps we retain in ack_queue. That limit should be something like 64, so that the largest possible size of an ACK frame would be smaller than the minimum MTU.

**Comment by sknat** [comment]
Hi @kazuho thanks for you reply, and sorry for the delay, 

You're right the issue happens while sending a lot of data on many streams with almost nothing sent back from the peer (I'm testing half-duplex). Capping the `ack_queue` such its content always fits inside one packet should solve the issue. 

I'll close this issue as it's part from #216

**Comment by kazuho** [comment]
I take back my argument that this is a bug caused by `ack_queue` becoming too large. In `record_receipt` function, we already limit the maximum number of ACK blocks we retain to 63 (QUICLY_ENCODE_ACK_MAX_BLOCKS). Therefore, the maximum size of the ACK packet we might ever try to emit is capped to below MTU.

And rereading the code, I think there is at least one issue in the `send_ack` function. The function has the following code:
```c
    /* emit ack frame */
Emit:
    if ((ret = allocate_frame(conn, s, QUICLY_ACK_FRAME_CAPACITY)) != 0)
        return ret;
    uint8_t *new_dst = quicly_encode_ack_frame(s->dst, s->dst_end, &space->ack_queue, ack_delay);
    if (new_dst == NULL) {
        /* no space, retry with new MTU-sized packet */
        if ((ret = commit_send_packet(conn, s, 0)) != 0)
            return ret;
        goto Emit;
    }
```

As stated in my previous comment, the assumption behind this code is that there shouldn't be a case where `allocate_frame` allocates a new QUIC *packet* and then we fail in encoding the ACK frame, because the size of the newly allocated QUIC packet is MTU bytes.

I do not think that is always true. When coalesced packet is being used, a new packet allocated by the `allocate_frame` function can be much smaller. In that case, we might see `quicly_encode_ack_frame` failing to write to the packet being allocated. Then, we'd call `commit_send_packet` that fails due to the assertion stated above, as we have written nothing to the packet yet.

I will write a patch for the problem.

**Comment by sknat** [comment]
Oh right, nice catch. I missed the coalesced case we going through it.
Thanks !

**Comment by kazuho** [comment]
So #243 should have fixed https://github.com/h2o/quicly/issues/240#issuecomment-567364878. @sknat Please let us know if you see the issue happen again. Thank you in advance.

**Comment by sknat** [comment]
Thanks a lot for the fix @kazuho 
I'll give it a go and ping you if the issue happens again.
